File: solutions/top-k-frequent-elements/top-k-frequent-elements.go
// @Title: Top K Frequent Elements
// @Author: colinjxc
// @Date: 2021-05-20T14:18:55+08:00
// @URL: https://leetcode-cn.com/problems/top-k-frequent-elements


func topKFrequent(nums []int, k int) []int {
	var res []int
	pq := PriorityQueue{}
	heap.Init(&pq)

	m := make(map[int]int)
	for _, n := range nums {
		m[n]++
	}

	for k, v := range m {
		heap.Push(&pq, &Item{
			key:   k,
			value: v,
		})
	}

	for len(res) != k {
		t := heap.Pop(&pq)
		if t == nil {
			break
		}
		item := t.(*Item)
		res = append(res, item.key)
	}
	return res
}

type Item struct {
	key   int
	value int
}

// PriorityQueue 实现堆接口
type PriorityQueue []*Item

func (p *PriorityQueue) Push(x interface{}) {
	*p = append(*p, x.(*Item))
}

func (p *PriorityQueue) Pop() interface{} {
	l := len(*p)
	if l == 0 {
		return nil
	}
	item := (*p)[l-1]
	*p = (*p)[:l-1]
	return item
}

func (p PriorityQueue) Len() int {
	return len(p)
}

func (p PriorityQueue) Swap(i, j int) {
	p[i], p[j] = p[j], p[i]
}

func (p PriorityQueue) Less(i, j int) bool {
	return p[i].value > p[j].value
}

File: solutions/top-k-frequent-elements/top-k-frequent-elements.cpp
// @Title: Top K Frequent Elements
// @Author: colinjxc
// @Date: 2022-02-06T23:16:47+08:00
// @URL: https://leetcode-cn.com/problems/top-k-frequent-elements


class Solution {
public:
    void qsort(vector<pair<int, int>>& v, int start, int end, vector<int>& ret, int k) {
        int picked = rand() % (end - start + 1) + start;
        swap(v[picked], v[start]);

        int pivot = v[start].second;
        int index = start;
        for (int i = start + 1; i <= end; i++) {
            if (v[i].second >= pivot) {
                swap(v[++index], v[i]);
            }
        }
        swap(v[start], v[index]);

        if (k <= index - start) {
            // have TOP(index-start+1), big then K
            qsort(v, start, index - 1, ret, k);
        } else {
            // have TOP(index-start+1), but small than K, take them all
            for (int i = start; i <= index; i++) {
                ret.push_back(v[i].first);
            }
            if (k > index - start + 1) {
                qsort(v, index + 1, end, ret, k - (index - start + 1));
            }
        }
    }

    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> freqs;
        for (auto& v: nums) {
            freqs[v]++;
        }

        vector<int> ret;
        vector<pair<int, int>> values(freqs.begin(), freqs.end());
        qsort(values, 0, values.size() - 1, ret, k);
        return ret;
    }
};

